src/System/Taffybar/WorkspaceHUD.hs

```diff
@@ -0,0 +1,504 @@
+                                    , contentsWorkspace =
+                                      ws { windowIds = []
+                                         , workspaceName = workspaceName ws ++ "fake"
+                                         }
+                                    , contentsConfig = cfg
+                                    }
+  WWC <$> updateWidget tempController ws
+
+instance WorkspaceWidgetController WorkspaceContentsController where
+  getWidget cc = Gtk.toWidget $ container cc
+  updateWidget cc newWorkspace = do
+    let currentWorkspace = contentsWorkspace cc
+        cfg = contentsConfig cc
+
+    when ((workspaceName currentWorkspace) /= (workspaceName newWorkspace)) $
+         Gtk.labelSetMarkup (label cc) (workspaceName newWorkspace)
+
+    newImages <-
+      if ((windowIds currentWorkspace) /= (windowIds newWorkspace) ||
+          (urgentIds currentWorkspace) /= (urgentIds newWorkspace))
+      then
+        updateImages cc newWorkspace
+      else
+        return $ iconImages cc
+
+    Gtk.widgetSetName (container cc) $ getWidgetName newWorkspace "contents"
+
+    maybe (return ()) (updateMinSize $ Gtk.toWidget $ container cc) $
+          minWSWidgetSize cfg
+
+    return cc { contentsWorkspace = newWorkspace
+              , iconImages = newImages
+              }
+
+updateMinSize :: Gtk.Widget -> Int  -> IO ()
+updateMinSize widget minWidth = do
+  W.widgetSetSizeRequest widget (-1) (-1)
+  W.Requisition w _ <- W.widgetSizeRequest widget
+  when (w < minWidth) $ W.widgetSetSizeRequest widget minWidth  $ -1
+
+defaultGetIconInfo :: WorkspaceHUDConfig -> X11Window -> IO IconInfo
+defaultGetIconInfo cfg w = do
+  -- TODO: handle custom files
+  icons <- withDefaultCtx $ getWindowIcons w
+  return $ if (null icons)
+           then IINone
+           else IIEWMH $ selectEWMHIcon (windowIconSize cfg) icons
+
+windowTitleClassIconGetter :: Bool -> (String -> String -> IconInfo) ->
+                              (WorkspaceHUDConfig -> X11Window -> IO IconInfo)
+windowTitleClassIconGetter preferCustom customIconF = fn
+    where fn cfg w = do
+            wTitle <- withDefaultCtx $ getWindowTitle w
+            wClass <- withDefaultCtx $ getWindowClass w
+            let customResult = customIconF wTitle wClass
+            defaultResult <- defaultGetIconInfo cfg w
+            let first = if preferCustom then customResult else defaultResult
+            let second = if preferCustom then defaultResult else customResult
+            return $ case first of
+                       IINone -> second
+                       _ -> first
+
+forkM :: Monad m => (i -> m a) -> (i -> m b) -> i -> m (a, b)
+forkM = ((liftM2 . liftM2) (,))
+
+updateImages :: WorkspaceContentsController -> Workspace -> IO [IconWidget]
+updateImages wcc ws = do
+  let cfg = contentsConfig wcc
+  iconInfos_ <- mapM (forkM (getIconInfo cfg $ cfg) return) $ windowIds ws
+  -- XXX: Only one of the two things being zipped can be an infinite list, which
```

Comment:
haha, there must be a better way..
code is a little too dense for me to come up with one quickly, tho, so i cant put my money where mouths are.